File: runtime/typescript/src/Nat.ts
import { Codec } from "./Codec.js";
import { FormatReader, FormatWriter } from "./FormatIO.js";
import { encodeVLQ, decodeVLQ } from "./VLQ.js";

export type Nat = bigint;

export const codec: Codec<Nat> = {
    read(reader: FormatReader): Promise<Nat> {
        return decodeVLQ(reader, false);
    },

    write(writer: FormatWriter, value: Nat): Promise<void> {
        return encodeVLQ(writer, false, value);
    },
};

export function fromInteger(n: bigint): Nat {
    return n;
}

File: runtime/java/src/main/java/dev/argon/verilization/runtime/Nat.java
package dev.argon.verilization.runtime;

import java.math.BigInteger;
import java.io.IOException;

public final class Nat {
    private Nat() {}

    public static BigInteger fromInteger(int i) {
        return BigInteger.valueOf(i).abs();
    }

    public static BigInteger fromInteger(long l) {
        return BigInteger.valueOf(l).abs();
    }

    public static BigInteger fromInteger(BigInteger i) {
        return i.abs();
    }

    public static final Codec<BigInteger> codec = new Codec<BigInteger>() {
        @Override
        public BigInteger read(FormatReader reader) throws IOException {
            return VLQ.decodeVLQ(reader, false);
        }

        @Override
        public void write(FormatWriter writer, BigInteger value) throws IOException {
            VLQ.encodeVLQ(writer, false, value);
        }

    };
}
